### DISCUSSION TITLE
[Contract Verification] Multiple issues with the era scan (sepolia) backend and having non-matching bytecode

### DISCUSSION THREAD

**Discussion by jordaniza** [zkSync-Community-Hub/zksync-developers#512] [Contract Verification] [answered]
### Environment

Testnet

### Issue Type

- [X] Plugin
- [X] Block Explorer

### Contract Address

0x0E50DA17F263e165780d103Dc20B4bB911146c67

### Compiler Type

Single file

### zkSolc Version

1.4.1

### Solc Version

0.8.17

### Contract Name

DAO

### Contract Code

https://gist.github.com/jordaniza/6fdce883e89f05fb0757d5dcdb9605bb

### Constructor Arguments

0x

### Hardhat Verify Plugin Version

"@matterlabs/hardhat-zksync-verify": "^0.6.0"
"@nomicfoundation/hardhat-verify": "^1.0.4"

### Repo Link (Optional)

https://github.com/aragon/osx/blob/zksync-testing/packages/contracts/src/core/dao/DAO.sol
https://github.com/aragon/osx/blob/zksync-testing/packages/contracts/src/framework/dao/DAOFactory.sol

### Additional Details

Hey team,

Having some real difficulties with verifying a couple contracts and think I'm outta options:

- When pinging the verification endpoint after running the verify command, I get an error which indicates that there is some discrepancy between:
  - my on-chain bytecode
  - the bytecode generated when sent by `hardhat-verify` to the verification endpoint
- This only happens for a handful of contracts like `DAOFactory.sol` and `DAO.sol` (and is successful for a number of others)
- I tried to flatten the `DAOFactory.sol` contract, this can't be done because of cyclic deps
- I was successfully able to flatten `DAO.sol` (linked) but got the same error when trying to manually upload to erascan.
- I got a different error when trying to use the plugin with `DAO.sol`: it takes no arguments in the constructor but runs into `Backend verification error: Constructor arguments are not correct` when passing `0x` as the `constructor args` **(EDIT: this is a known bug)**
- We can't see the sepola erascan backend, so we have no idea what specifically is causing the discrepancy.

I've added full details in [this thread](https://gist.github.com/novaknole/450e4614b37517d6eb72956dba196ea0?permalink_comment_id=5049313#gistcomment-5049313)

Would really welcome a hand here as we are blocked trying to deploy until we resolve this and I am really confused as to what the problem might be - thanks in advance!

**Comment by bxpana** [comment]
Appreciate the post and the repo link. Will check things out to see what's going on.

**Comment by kiriyaga-txfusion** [comment, marked as answer]
Hi [@jordaniza](https://github.com/jordaniza),
After our investigation, here are some key points:
- When resolving contracts needed for DAOFactory verification hardhat try to only gather dependent contracts but this context wasn't enough for zksolc compiler on BE verification side. Once we sent full context to the BE, verification succeeded.
- Beside this, since the code base needed for DAOFactory verification is large, hardhat verify process ended so the fallback mechanism that send full context is not triggered. Nonetheless, contract DAOFactory is now verified, and you can access it [here](https://sepolia.explorer.zksync.io/address/0x6Ec0D827c95D3A0768BB34f59eA56E2b18D77AbD).
- Regarding DAO verification, currently the backend verification system does not support verifying contracts deployed from another contract. We are working to support this feature and we will let you know once this is implemented.

**Reply by jordaniza** [reply]
In case anyone stumbles across this: there are some limitations on the backend verifier which the ZkSync team were helpful enough to resolve for us. Recommend reaching out to them in the short term
